### DISCUSSION TITLE
Questions about the FIFO depth between layers

### DISCUSSION THREAD

**Discussion by wangtianxing1991** [Xilinx/finn#383] [Q&A] [answered]
Quick summary
Why the FIFO depth between layers affects the speed

Details
FINN uses the “step_set_fifo_depths” method to automatically set the FIFO depth between layers. I tried to reduce the FIFO depth (from 128 to 16) and found that the network was running slower on FPGA. In my opinion, when fclayer1 has less running cycles than fclayer2, that is, fclayer1 runs faster. No matter how deep the FIFO is between them, it will be full. Then fclayer1 needs to wait until FIFO is not full before continuing to write. In my opinion, in addition to the short time the model has just run, FIFO runs in this state most of the time, that is, "FIFO full - FIFO nearly full - FIFO full - FIFO nearly full...". So I think the FIFO depth does not affect the speed. I know there must be something wrong with my idea. I don't know where I didn't take it into account. Please help me, thank you.

An automatic FIFO depth setting method is mentioned in Gitter, as shown in the figure below. "Pushing in a steady stream of inputs to fill up the pipeline" is mentioned. I want to know how much input data is needed? thank you.

![image](https://user-images.githubusercontent.com/79840616/136512060-2b66fd50-5759-4aa3-83ec-4fd08dc7d240.png)

![image](https://user-images.githubusercontent.com/79840616/136510269-e9aba4ea-7c7b-43a5-8f16-63cf87ecb42e.png)

**Comment by heborras** [comment, marked as answer]
Hi @wangtianxing1991,
thank you for your question. This question is actually quite complicated to understand, and very much not trivial, so the question is highly appreciated.

I'll answer your question in two sections.

### First quesiton
I hope it is correct if I reformulate the first part of your question as: **"Why do small FIFOs hurt performance?"**
This question is, in my understanding, basically equivalent to: **"Why do we need FIFOs?"** Because if small FIFOs wouldn't hurt performance, then why have FIFOs at all.

**The TL;DR is: The `StreamingFCLayer_Batch` layers read and write data in bursts, which leads to stalling layers if the FIFOs are too small.**

The long answer with an example is this:

The answer to this lies in how the `StreamingFCLayer_Batch` node operates. The `StreamingFCLayer_Batch` node encapsulates this function from the finn-hls library: https://github.com/Xilinx/finn-hlslib/blob/1ceb9094a5fbb8d0a4a9064a16c0ee2073cb8a33/fclayer.h#L94
It is somewhat difficult to see, but the inputs and outputs of this functions are bursty. Meaning that the function doesn't output data at a constant rate, but sometimes more samples at a time, then less for a while, then again more and so on. The same is true for the read operations.

Okay, let's make up an example here for illustration. The example is somewhat extreme, but mirrors reality.
Imagine we have FC1, which writes one sample every four cycles and FC2, which reads one sample every five cycles. So FC1 writes faster than FC2 can read.
In an ideal world we would cache one sample from the FC1 output and always have something for FC2 to work with. This would mean that FC2 would never stall and FC1 would stall sometimes. And the throughput would be dictated by how fast FC2 can run, since it is the slowest element in our pipeline.
The ideal case is illustrated below:

| Cycle | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 |
|-------|---|---|---|---|---|---|---|---|---|----|----|----|
| FC1   | X | W | - | - | - | X | W | - | - | -  | X  | W  |
| Cache | 1 | 0 | 1 | 1 | 1 | 1 | 0 | 1 | 1 | 1  | 1  | 0  |
| FC2   | R | - | - | - | - | R | - | - | - | -  | R  | -  |

For FC1 "W" is a write operation, for FC2 "R" is a read operation and for both "-" is a cycle in which computation is happening and "X" is a cycle in which the layer has stalled.
Our cache in between is either filled with data for FC2, denoted by a "1" or empty, denoted by a "0".
Read and writes to the cache take one cycle to apply. We also assume that we start with a full cache.
And here you can see what you would intuitively expect: FC2 never stalls and FC1 stalls sometimes.

Now we add that FC2 doesn't read at a constant rate, but reads in bursts. We assume the layer reads three elements after another, followed by 12 cycles of calculation and no reads. Meaning that on average we still read one sample every five cycles. And we would end up with a read pattern like this, if we always have enough data to read:
| Cycle | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 |
|-------|---|---|---|---|---|---|---|---|---|----|----|----|----|----|----|----|----|----|----|----|----|----|
| FC2   | R | R | R | - | - | - | - | - | - | -  | -  | -  | -  | -  | -  | R  | R  | R  | -  | -  | -  | -  |

Now lets add FC1 and the cache on top of this:
| Cycle | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 |
|-------|---|---|---|---|---|---|---|---|---|----|----|----|----|----|----|----|----|----|----|----|----|----|
| FC1   | X | W | - | - | - | W | - | - | - | W  | X  | X  | X  | X  | X  | X  | X  | X  | X  | X  | W  | -  |
| Cache | 1 | 0 | 1 | 0 | 0 | 0 | 1 | 0 | 0 | 0  | 1  | 1  | 1  | 1  | 1  | 1  | 1  | 1  | 1  | 1  | 0  | 1  |
| FC2   | R | X | R | X | X | X | R | - | - | -  | -  | -  | -  | -  | -  | -  | -  | -  | -  | R  | X  | R  |

Wow, what has happened? Suddenly FC2 is stalling significantly. And because FC2 stalls it is even slower than before and since it already was the slowest element in the pipeline, the throughput of the whole network also drops.
Well the reason that FC2 is stalling is that the cache can't supply enough data to satisfy the burst of three samples, which FC2 needs to not stall.

Now let's convert the cache into a FIFO, which can store two elements. This way we should be able to store enough samples to give FC2 the three samples, when it needs them:

| Cycle | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 |
|-------|---|---|---|---|---|---|---|---|---|----|----|----|----|----|----|----|----|----|----|----|----|----|
| FC1   | X | W | - | - | - | W | - | - | - | W  | -  | -  | -  | X  | X  | X  | W  | -  | -  | -  | W  | -  |
| FIFO  | 2 | 1 | 1 | 0 | 0 | 0 | 1 | 1 | 1 | 1  | 2  | 2  | 2  | 2  | 2  | 2  | 1  | 1  | 0  | 0  | 0  | 1  |
| FC2   | R | R | R | - | - | - | - | - | - | -  | -  | -  | -  | -  | -  | R  | R  | R  | -  | -  | -  | -  |

And indeed, this helps: FC2 doesn't stall anymore and FC1 stalls sometimes.

This is a small example, but one could extend it to having FC2 read more than three samples at a time. Then one could also see that too small FIFOs gradually diminish the throughput as they get smaller.

The same issue with stalling layers is true when a layer writes in bursts. You could write that down in the same way, which I just did for reading in bursts.

And the whole thing gets almost unpredictably complicated when the layers both read and write in bursts.
This complicated behavior is also the reason why we need to actually simulate the network, because so far we weren't able to come up with a good estimate for the FIFO depths in such a chaotic system.

And that's the reason why we need FIFOs. Basically to cache enough data to satisfy the burst read/writes of the HLS layers.

@maltanar: Please correct me if I explained something incorrectly here.


### Second question

> An automatic FIFO depth setting method is mentioned in Gitter, as shown in the figure below. "Pushing in a steady stream of inputs to fill up the pipeline" is mentioned. I want to know how much input data is needed? 

The question boils down to: **"How long do we need to run the the simulation until the whole pipeline is running at full capacity?"**
And the source code has an answer for you. This is for how many cycles the simulation is run in FINN: https://github.com/Xilinx/finn/blob/cbdfcfc3773e6b7cd0ceb81b0088aabf70065f38/src/finn/transformation/fpgadataflow/set_fifo_depths.py#L303
It's basically the number of cycles it takes one image to travel through the whole network plus the number of cycles it takes one image to be consumed at the input of the network.



I hope this answers your questions. If not, don't hesitate to ask a follow up question.

**Reply by wangtianxing1991** [reply]
Thanks Hendrik. Your answer is very clear and detailed. This made me fully understand. After reading your answer, I thought for a long time, why "FC2 doesn't read at a constant rate". After reading the Finn-HLS code, I understand that each layer should prepare enough data before starting the calculation, so layers read in bursts.

**Reply by maltanar** [reply]
Excellent write-up @HenniOVP ! I will add one additional point for future reference: a linear topology can still work (although slower than the designed performance) with too-small FIFOs, but for branching topologies like ResNets FIFOs can be essential to avoid deadlock between branches.

**Comment by giovannipollo** [comment]
Thanks for the clear explanation. Following this topic, what would be the "best way" to compute the correct depth of the FIFOs given a specific design with a particular number of PE and SIMD per layer? 
For example, are the `folding_config.json` present in the models of the `finn-examples` repository (https://github.com/Xilinx/finn-examples/tree/main/build) obtained with a specific pattern? If yes, what is the pattern?
Thanks in advance!

**Reply by heborras** [reply]
Hi @giop98 
the "best way" to set the FIFO sizes depends on what your needs are.

If you are looking for the highest guaranteed performance during inference, then having FINN run the full RTL simulation and determine the FIFO sizes is likely going to get you there. This is basically what my answer above is referencing. However, determining the FIFO sizes this way may take very long for large networks and it may also lead to larger than required FIFOs.

Since v0.9 there is also a new way to run this simulation in parallel for linear network typologies. See #689  for more details on how it works and how to use it. This method will run significantly faster on multi-core systems and is likely to produce smaller FIFOs for deep networks. However, as mentioned, it will only work for non-branching network architectures.

Now these two approaches are fine if all you are looking for is a well performing implementation, you are happy with the synthesis times and everything fits onto your FPGA.

If however you don't want to run the RTL simulation every time you build an accelerator with FINN, then you should look into fixing the FIFO sizes ahead of time with the `folding_config.json` you mentioned before. With this you can effectively skip the RTL simulation and tell FINN to just use those values instead.

A similar approach is also required if the generated accelerator doesn't fit onto your FPGA out of the box. What I would recommend doing in that case is: Having FINN run the RTL simulation to determine an optimal staring point for the FIFO sizes and then reducing the FIFOs manually with a custom `folding_config.json`. Finding good FIFO sizes here is not easy and will take quite some time. In general and in my experience, the slowdown incurred by too small FIFOs is roughly linear. So you should keep this in mind when reducing the FIFO sizes by hand. Depending on what your time budget is it may also be useful to run a limited parameter sweep for different FIFO settings to get an overview of how the FINN generated accelerator reacts to different FIFO sizes.

As for how the `folding_config.json` files in `finn-examples` repository came about I can't really answer, since I wasn't involved in the development of any of the models, which use them. My guess is that they are manually tuned for almost all models. Especially the `bnn-pynq` models predate the introduction of automatic FIFO sizing in FINN, so these are most certainly hand tuned. As for the other models: Here especially the larger models might have required extensive hand tuning to to get them to run on a given FPGA.
